lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,662 @@
+ * 2. Generate an img tag enabling the browser to render the image without the AMP runtime being loaded.
+ *
+ * Hero images are either identified automatically or can be explicitly defined by adding an `data-hero` attribute to
+ * the element.
+ *
+ * This transformer supports the following options:
+ *
+ * * `preloadHeroImage`: [true|false] - enables or disables hero image preloading. The default is `true`.
+ *
+ * This is ported from the NodeJS optimizer.
+ *
+ * @version 3429af9d91e2c9efe1af85757499e5a308755f5f
+ * @link    https://github.com/ampproject/amp-toolbox/blob/3429af9d91e2c9efe1af85757499e5a308755f5f/packages/optimizer/lib/transformers/PreloadHeroImage.js
+ *
+ * @package ampproject/optimizer
+ */
+final class PreloadHeroImage implements Transformer
+{
+
+    /**
+     * Class(es) to apply to a serverside-rendered image element.
+     *
+     * @var string
+     */
+    const SSR_IMAGE_CLASS = 'i-amphtml-fill-content i-amphtml-replaced-content';
+
+    /**
+     * List of attributes to copy onto an SSR'ed image.
+     *
+     * @var string[]
+     */
+    const ATTRIBUTES_TO_COPY = [
+        Attribute::ALT,
+        Attribute::ATTRIBUTION,
+        Attribute::OBJECT_FIT,
+        Attribute::OBJECT_POSITION,
+        Attribute::REFERRERPOLICY,
+        Attribute::SRC,
+        Attribute::SRCSET,
+        Attribute::SIZES,
+        Attribute::TITLE,
+    ];
+
+    /**
+     * Maximum number of hero images defined via data-hero attribute.
+     *
+     * @var int
+     */
+    const DATA_HERO_MAX = 2;
+
+    /**
+     * List of AMP elements that are an embed that can have a placeholder.
+     *
+     * The array has values assigned so that we can do a fast hash lookup on the element name.
+     *
+     * @var bool[]
+     */
+    const AMP_EMBEDS = [
+        Extension::AD            => true,
+        Extension::ANIM          => true,
+        Extension::BRIGHTCOVE    => true,
+        Extension::DAILYMOTION   => true,
+        Extension::FACEBOOK      => true,
+        Extension::GFYCAT        => true,
+        Extension::IFRAME        => true,
+        Extension::IMGUR         => true,
+        Extension::INSTAGRAM     => true,
+        Extension::PINTEREST     => true,
+        Extension::REDDIT        => true,
+        Extension::TWITTER       => true,
+        Extension::VIDEO         => true,
+        Extension::VIDEO_IFRAME  => true,
+        Extension::VIMEO         => true,
+        Extension::WISTIA_PLAYER => true,
+        Extension::YOUTUBE       => true,
+    ];
+
+    /**
+     * XPath query to relatively fetch all noscript > img elements.
+     *
+     * @var string
+     */
+    const NOSCRIPT_IMG_XPATH_QUERY             = './/noscript[ img ]';
+
+    /**
+     * Regular expression pattern to extract the URL from a CSS background-image property.
+     *
+     * @var string
+     */
+    const CSS_BACKGROUND_IMAGE_URL_REGEX_PATTERN = '/background-image\s*:\s*url\(\s*(?<url>[^)]*\s*)/i';
```

Comment:
No need to match the trailing `\s*` right?

```suggestion
    const CSS_BACKGROUND_IMAGE_URL_REGEX_PATTERN = '/background-image\s*:\s*url\(\s*(?<url>[^)]*)/i';
```

Reply:
The trailing space was meant to go between the matched URL and the closing parentheses, but the latter are missing. This would be a better regex:
```suggestion
    const CSS_BACKGROUND_IMAGE_URL_REGEX_PATTERN = '/background-image\s*:\s*url\(\s*(?<url>[^)]*?)\s*\)/i';
```